init.lua

```diff
@@ -510,56 +496,49 @@ end
 
 local function get_filter_items(data, player)
 	local filter = data.filter
-	if searches[filter] then
-		data.items = searches[filter]
-		return
-	end
-
-	local items_list = progressive_mode and data.init_filter_items or init_items
+	local items_list = init_items
 	local inv = player:get_inventory()
 	local filtered_list, c = {}, 0
 
 	for i = 1, #items_list do
 		local item = items_list[i]
 		local item_desc = reg_items[item].description:lower()
 
-		if filter ~= "" then
-			if item:find(filter, 1, true) or item_desc:find(filter, 1, true) then
-				c = c + 1
-				filtered_list[c] = item
+		if (filter == "" or item:find(filter, 1, true) or item_desc:find(filter, 1, true)) then
+			local _recipes, has_item
+			if progressive_mode then
+				_recipes, has_item = recipe_in_inv(inv, item)
+			else
+				has_item = true
 			end
-		elseif progressive_mode then
-			local _, has_item = recipe_in_inv(inv, item)
 			if has_item then
 				c = c + 1
 				filtered_list[c] = item
 			end
 		end
 	end
-
-	if progressive_mode then
-		if not data.items then
-			data.init_filter_items = filtered_list
-		end
-	elseif filter ~= "" then
-		-- Cache the results only if searched 2 times
-		if searches[filter] == nil then
-			searches[filter] = false
-		else
-			searches[filter] = filtered_list
-		end
-	end
-
 	data.items = filtered_list
 end
 
-local function init_data(user, name)
-	player_data[name] = {filter = "", pagenum = 1, iX = sfinv_only and 8 or DEFAULT_SIZE}
-	if progressive_mode then
-		get_filter_items(player_data[name], user)
+local show_fs = function(player, player_name)
+	local data = player_data[player_name]
+	if not data.items then
+		get_filter_items(data, player)
```

Comment:
If the guide is cleared you force it to rebuild a list while before it was cached. I can't see that as an optimization.

Reply:
get_filtered_items does not rebuild the initial list, just apply searches. I reimplemented the caching of string-search results again inside the get_filter_items()